crates/runestick-macros/src/context.rs

```diff
@@ -487,61 +492,84 @@ impl Context {
         &self,
         ident: T,
         name: &TokenStream,
-        install_with: &TokenStream,
+        installers: &TokenStream,
         tokens: &Tokens,
+        generics: &syn::Generics,
     ) -> Result<TokenStream, Vec<syn::Error>>
     where
         T: Copy + ToTokens,
     {
-        let any = &tokens.any;
-        let context_error = &tokens.context_error;
-        let hash = &tokens.hash;
-        let module = &tokens.module;
-        let named = &tokens.named;
-        let pointer_guard = &tokens.pointer_guard;
-        let raw_into_mut = &tokens.raw_into_mut;
-        let raw_into_ref = &tokens.raw_into_ref;
-        let raw_str = &tokens.raw_str;
-        let shared = &tokens.shared;
-        let type_info = &tokens.type_info;
-        let type_of = &tokens.type_of;
-        let unsafe_from_value = &tokens.unsafe_from_value;
-        let unsafe_to_value = &tokens.unsafe_to_value;
-        let value = &tokens.value;
-        let vm_error = &tokens.vm_error;
-        let install_into_trait = &tokens.install_with;
-
+        let Tokens {
+            any,
+            context_error,
+            hash,
+            module,
+            named,
+            pointer_guard,
+            raw_into_mut,
+            raw_into_ref,
+            raw_str,
+            shared,
+            type_info,
+            type_of,
+            unsafe_from_value,
+            unsafe_to_value,
+            value,
+            vm_error,
+            install_with,
+            ..
+        } = &tokens;
+
+        let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+        let generic_names = generics.type_params().map(|v| &v.ident).collect::<Vec<_>>();
+
+        let impl_named = if !generic_names.is_empty() {
+            quote! {
+                impl #impl_generics #named for #ident #ty_generics #where_clause {
+                    const NAME: #raw_str  = #raw_str::from_str(#name);
+
+                    fn exact() -> String {
```

Comment:
`fn full_name` perhaps? Conversely we can rename the constant `BASE_NAME`.

It would also be useful to support some form of formatting protocol for `full_name` (taking `buf: &mut String` would be the easiest), since otherwise each recursion into `exact` could end up allocating while it doesn't really have to.

Reply:
Part 1: Sure!

The second one sounds like premature optimization for me, but I'll do it if you insist :-) As it is right now, it's only used in the Module::ty call, which is once per application lifetime and type. That's... Unlikely to be the performance bottleneck any time soon. ;-)